src/App.js

```diff
@@ -25,44 +24,97 @@ const generateSquares = () => {
-  //   Then pass it into the squares as a callback
+  const checkForWinner = (squares) => {
+    let x = 0
+    let o = 0
+    let empty = false
+    const possibilities = [
+      [squares[0][0].value, squares[0][1].value, squares[0][2].value],
+      [squares[1][0].value, squares[1][1].value, squares[1][2].value],
+      [squares[2][0].value, squares[2][1].value, squares[2][2].value],
+      [squares[0][0].value, squares[1][0].value, squares[2][0].value],
+      [squares[0][1].value, squares[1][1].value, squares[2][1].value],
+      [squares[0][2].value, squares[1][2].value, squares[2][2].value],
+      [squares[0][0].value, squares[1][1].value, squares[2][2].value],
+      [squares[0][2].value, squares[1][1].value, squares[2][0].value]
+    ]
 
+    for(let i = 0; i < possibilities.length; i++){
+      if(possibilities[i][0] === '' || possibilities[i][1] === '' || possibilities[i][2] === ''){
+        empty = true
+      }
+      if(possibilities[i][0] === possibilities[i][1] && possibilities[i][1] === possibilities[i][2]){
+        if(possibilities[i][0] === 'x'){
+          x += 1
+        }
+        if( possibilities[i][0] === 'o'){
+          o += 1
+        }
+      }
+    }
 
-  const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+    if(x === 0 && o === 0 && empty){
+      setWinner(null)
+    }
+    if(x === 0 && o === 0 && !empty)
+      setWinner('Tie')
+    if( x > o){
+      setWinner('x')
+    }
+    if(x < o){
+      setWinner('o')
+    }
+  }
 
+  const updateSquaresByID = (squareID) => {
+    for(let i = 0; i < squares.length; i++){
+      for(let j = 0; j < squares[i].length; j++){
+        if(squares[i][j].id === squareID && squares[i][j].value === ''){
+          squares[i][j].value = currentPlayer
+          updateCurrentPlayer(currentPlayer, setCurrentPlayer)
+          setSquares(squares)
+        }
+      }
+    }
   }
 
+  const onClickCallback = ((squareID) => {
+    if(winner === null){
+      updateSquaresByID(squareID)
+      checkForWinner(squares)
+    }
+  })
+
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares((generateSquares()))
+    setCurrentPlayer(PLAYER_1)
+    setWinner(null)
   }
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        {winner === 'Tie' ? <h2 className='tieGame'>The game is tied!</h2> : ''}
+        {winner === null ?  <h2>The current player is {currentPlayer}</h2> : ''}
+        {winner === 'o' || winner === 'x'? <h2>Winner is {winner.toLowerCase()}</h2> : ''}
```

Comment:
I would be inclined to pull this out of the markup here and into code that runs before the JSX return stuff, or into a helper function. There's sort of two things going on here: figuring out the `className`, and figuring out the `h2` contents. Calculating those ahead of time before getting here can help make that logic more visible, and make the markup here a little clearer.